### PR TITLE
feat(Popover): Support hover trigger

### LINKED ISSUE #135: Popover: Update to support the latest version of popper.js
Last I heard, there were some style changes that were affecting our component.

- [ ] Update to the latest popper.js library
- [ ] Fix any broken styles

### PR DESCRIPTION
**Pull request by atlwendy**
closes #135

### CODE REVIEW

File: .github/workflows/pr-ci.yml [hunk 1]
```diff
@@ -73,6 +73,9 @@ jobs:
         if: env.SHOULD_RUN == 'true'
       - name: Regression test on affected projects
         run: tools/ci/run-chromatic-with-changed-projects.sh
+        env:
```
Comment (benjamincharity) [inline .github/workflows/pr-ci.yml]: This should probably be a separate commit since it isn't really related to popover

File: libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts [hunk 2]
```diff
@@ -132,6 +136,12 @@ export class TsPopoverTriggerDirective implements OnInit, OnDestroy, OnChanges,
   }
   public _position = TsPopoverPositions.Bottom;
 
+  /**
+   * Set trigger type of popover, i.e. click or hover
+   */
+  @Input()
+  public showTrigger: TsTrigger = 'click';
```
Comment (benjamincharity) [inline libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: Hmm this name confuses me a bit as it sounds like the 'trigger' is what will be shown. Maybe just `trigger`? Or `popoverTrigger`? or `triggerEvent` etc?

File: libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts [hunk 3]
```diff
@@ -200,15 +210,36 @@ export class TsPopoverTriggerDirective implements OnInit, OnDestroy, OnChanges,
...
-    // eslint-disable-next-line deprecation/deprecation
-    (fromEvent<MouseEvent>(this.elementRef.nativeElement, 'click'))
-      .pipe(
-        untilComponentDestroyed(this),
-      ).subscribe(event => {
-        this.toggle();
-        event.stopPropagation();
-        event.preventDefault();
-      });
+    switch (this.showTrigger) {
```
Comment (benjamincharity) [inline libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: Seems like it might be better to have the setter for the `showTrigger` input wire this up?

File: specs/ui-popover/popover-trigger.directive.spec.ts [hunk 4]
```diff
@@ -70,15 +72,15 @@ describe(`popover trigger`, () => {
 
     describe(`ID`, () => {
       test(`should support a custom ID`, fakeAsync(() => {
-        setup(Basic);
+        setup(testComponents.Basic);
```
Comment (benjamincharity) [inline specs/ui-popover/popover-trigger.directive.spec.ts]: Is there a reason to go back to this format? If we just use `Basic` the IDE will handle the import for us

File: specs/ui-popover/test-components.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
+import { CommonModule } from '@angular/common';
+import { Component, NgModule } from '@angular/core';
+
+import { TsPopoverModule, TsPopoverPosition } from '@terminus/ui-popover';
```
Comment (benjamincharity) [inline specs/ui-popover/test-components.ts]: Formatting. 

also, fyi there is a webstorm formatting setting that will pop them on their own lines as they are added

File: libs/ui/popover/.storybook/main.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  addons: ['@storybook/addon-knobs/register']
```
Comment (benjamincharity) [inline libs/ui/popover/.storybook/main.js:2]: I wonder if we can extend a base config like eslint/etc for things like this that we know everyone will need..

File: libs/ui/button/.storybook/main.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
```
Comment (benjamincharity) [inline libs/ui/button/.storybook/main.js]: This also seems like it's with the wrong commit..

File: apps/showcase-ui/src/app/components/popover/popover.component.html [hunk 8]
```diff
@@ -1,26 +1,65 @@
...
-    style="margin: 200px 250px;"
-    [id]="myId"
-    (popoverOnShown)="popoverOnShown($event)"
-    (popoverOnHidden)="popoverOnHidden($event)"
-  >Click me!</button>
-</div>
+<div fxLayout="row" style="display: flex">
+
+  <div style="max-width: 45%; height: 400px; flex: 1 1 100%; " class="demo-popover">
+    <span ><h4 style="margin-left: 10px">With Click Trigger</h4></span>
```
Comment (benjamincharity) [inline apps/showcase-ui/src/app/components/popover/popover.component.html]: Is this span needed?

**Review by benjamincharity — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-ci.yml (hunk 1 above)
Comment (atlwendy) [reply .github/workflows/pr-ci.yml]: True. I was merely using this pr opportunity to see how chromatic behaves. For a branch that only changes chromatic setting, nothing would get run (because no affected libs returned)

File: libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts (hunk 2 above)
Comment (atlwendy) [reply libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: Yeah it's meant to mean how trigger shown. Thought about `trigger` but felt it was generic. `popoverTrigger` might be a better choice.

File: libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts (hunk 3 above)
Comment (atlwendy) [reply libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: Hmm... what'd be the main difference for it being here or in setter? My thought was that the event wiring up should be happening when component starts up.

File: specs/ui-popover/popover-trigger.directive.spec.ts (hunk 4 above)
Comment (atlwendy) [reply specs/ui-popover/popover-trigger.directive.spec.ts]: ah no, I think I used one example (input probably) and it might just have out of dated format..

File: .github/workflows/pr-ci.yml (hunk 1 above)
Comment (benjamincharity) [reply .github/workflows/pr-ci.yml]: Yeah I have no problem with it being part of this PR, just not the commit 👍

File: libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts (hunk 3 above)
Comment (benjamincharity) [reply libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: I'm on board with that 👍

File: libs/ui/popover/README.md [hunk 9]
```diff
@@ -107,6 +108,22 @@ Defines where the popover is positioned, relative to current element:
 
 The default is `bottom`. For all available positions, please see the [popper.js placement docs][popper-placements].
 
+### Popover Trigger
```
Comment (benjamincharity) [inline libs/ui/popover/README.md:111]: Looks like we need to run `yarn run docs:toc` again

File: libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts [hunk 10]
```diff
@@ -200,15 +210,36 @@ export class TsPopoverTriggerDirective implements OnInit, OnDestroy, OnChanges,
...
-    // eslint-disable-next-line deprecation/deprecation
-    (fromEvent<MouseEvent>(this.elementRef.nativeElement, 'click'))
-      .pipe(
-        untilComponentDestroyed(this),
-      ).subscribe(event => {
-        this.toggle();
-        event.stopPropagation();
-        event.preventDefault();
-      });
+    switch (this.popoverTrigger) {
```
Comment (benjamincharity) [inline libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: huh the old comments seem to have vanished.. This is definitely style vs functional imo but the general consensus in Angular seems to be that we should keep logic out of `ngOnInit` when possible (just like we would with the constructor) -of course there are exceptions where it's needed for some reason.

In this case I also think it's probably a safer approach to add it to the setter since we have no guarantee that the consumer's input will be set prior to the component being initialized. So if they're waiting on some async data (since everything comes from the store/localstorage/server) the setting value could arrive after initialization.

If, for instance, the logic here also required a second `@Input` to be set to work, then the logic should probably be called from `ngOnChanges`

**Review by benjamincharity — CHANGES_REQUESTED** [review]
mostly cleared up.. 2 tiny notes

File: libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts (hunk 10 above)
Comment (atlwendy) [reply libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: Makes sense! My main concern was that this event subscription shouldn't be called more than once, because otherwise it'd get multiple subscriptions from mouse event and messes it up (that's why it definitely cannot be called from `ngOnChanges`). So as long as consumers aren't setting the trigger multiple times, it's all good.

Comment (benjamincharity) [reply libs/ui/popover/src/lib/trigger/popover-trigger.directive.ts]: yeah I think either way you'll have to 

a) check for existing sub
b) kill it if it exists
c) create new sub

iirc we have to do the same type of thing somewhere else... maybe input or cohort date?

File: libs/ui/popover/.storybook/main.js (hunk 6 above)
Comment (atlwendy) [reply libs/ui/popover/.storybook/main.js:2]: Yeah, will open a separate PR for all chromatic related tweaks..

**Review by benjamincharity — APPROVED** [review]
